### DISCUSSION TITLE
API for retrieving resource version meta for build outputs?

### DISCUSSION THREAD

**Discussion by jaredstehler** [concourse/concourse#6352] [General]
For integration with other systems (i.e. spinnaker), it's necessary to get the metadata for build outputs; currently the approach taken by spinnaker is to parse the entire build events stream and extract messages containing the "put" step output and "metadata" field contained therein. This is inefficient and slow (esp. for chatty builds).

It seems like the information I need is in the db, like:
```sql
concourse=> 
select versions.metadata 
from resource_config_versions versions, build_resource_config_version_outputs build_outputs 
where versions.version_md5=build_outputs.version_md5 
and build_outputs.build_id=?;
```

I don't believe this is retrievable via the existing API (as far as I can tell). `/api/v1/builds/:build/resources` only includes minimal info (i.e. `s3` put: 
```
    {
      "name": "ala-user-api-deploy-yaml",
      "version": {
        "path": "artifacts/ala-user-api/deploy-1.0.278-b5d970f.yaml"
      }
    },
```
 
Is there something I'm missing? Or any reason not to include the version metadata in the /resources output? Or perhaps even the resource version ID, so I could then use `/api/v1/teams/:team/pipelines/:pipeline/resources/:resource/versions/:versionId`?

**Comment by taylorsilva** [comment]
The API is not designed for that and we generally discourage people trying to use the API to automate things.

Would it be possible to instead `put` the metadata you need somewhere else or maybe generate your own and `put` that somewhere (s3?).

There is also some metadata you can grab in `get` and `put` steps: https://concourse-ci.org/implementing-resource-types.html#resource-metadata

Sorry if this misses the mark for you. I don't understand why you need that data specifically from your SQL query.

**Comment by jaredstehler** [comment]
in order to use concourse as a pipeline trigger and artifact source in spinnaker, we need to be able to retrieve the put outputs for a given build.

currently, this is done by:
1. retrieving the build plan (`/api/v1/builds/44076/plan`), and retrieving all get/put step ids:
```
                  {
                    "id": "5fd00a5c",
                    "on_success": {
                      "step": {
                        "id": "5fd00a5a",
                        "put": {
                          "type": "s3",
                          "name": "ala-user-api-deploy-yaml-latest",
                          "resource": "ala-user-api-deploy-yaml-latest"
                        }
                      },
                      "on_success": {
                        "id": "5fd00a5b",
                        "get": {
                          "type": "s3",
                          "name": "ala-user-api-deploy-yaml-latest",
                          "resource": "ala-user-api-deploy-yaml-latest"
                        }
                      }
                    }
                  },
```

2. streaming the events endpoint (`/api/v1/builds/44076/events`) pulling out everything with "metadata" and mapping by "origin.id", then merging get and put into consolidated dict:
```
...
id: 179
event: event
data: {"data":{"origin":{"id":"5fd00a5a"},"time":1607523690,"exit_status":0,"version":{"version_id":"PCxYH3uGiu5eOujI7Fo3xxx"},"metadata":[{"name":"filename","value":"deploy-latest.yaml"},{"name":"url","value":"https://s3-us-west-2.amazonaws.com/ci-artifacts-bucket/artifacts/ala-user-api/deploy-latest.yaml?versionId=PCxYH3uGiu5eOujI7Fo3txxx"}]},"event":"finish-put","version":"5.1"}
...
```

3. at this point, spinnaker is able to extract s3 artifacts from this build and make them available to the deployment pipeline.

For resources such as S3, the information I require is in that `url` metadata field. This information exists in that `resource_config_versions` table, but is oddly absent from the `/api/v1/builds/:id/resources` output.

To me, it would make sense to be able to provide a single endpoint for retrieving metadata about the actual resolved inputs and outputs for a given build.

Side-note; why the reluctance to have integrations directly use the API?

**Comment by taylorsilva** [comment]
My reply is kinda all over the place, I had too many thoughts haha. I really want to understand your problem and see what other solutions are available to besides relying on the API. Reply is in two parts. Possible concrete solution first, more why-asking in second part.

---

> For resources such as S3, the information I require is in that url metadata field.

Not 100% sure about this but I think the s3 resource _may_ give you all this info.
In the docs for the s3 resource I see that a `get` step will return you the `url` and file `version` which appears to match what is in the the `metadata` field: https://github.com/concourse/s3-resource/#in-fetch-an-object-from-the-bucket

After your job `puts` the artifact into s3 an implicit `get` step is ran. ([from the put step docs](https://concourse-ci.org/jobs.html#put-step))
After the put step runs you could take the url + version and push that somewhere for spinnaker/other system to pick up as the next new version. Pushing it somewhere doesn't have to be another s3 bucket, it could be literally anything, like a database, API enpoint, etc. YOU get to to decide what this looks like!

If the s3 is not giving you this info then maybe you could modify the s3 to actually return this info in the get step. Seems simpler than trying to wrestle with the API.

---

Sorry I don't know anything about spinnaker so it's not obvious to me why:
> .. it's necessary to get the metadata for build outputs..

Why is this metadata necessary? From my POV Spinnaker was not made with Concourse in mind so why does it care about the build metadata that Concourse generates? Could you not generate your own metadata while a task is running? Is uniqueness a  concern? I feel like you could generate something pretty unique based on git sha + datetime and push that somewhere for successful builds ([samver?](https://samver.org/))? Jobs and steps both have an `on_success` hook: https://concourse-ci.org/jobs.html#schema.job.on_success

Could you, on every successful build, push a file that contains the git sha + datetime (+ if you want metadata in a task https://github.com/swce/metadata-resource), and have spinnaker or your other build system use that?

In general, Concourse wants you to **externalize everything** related to state to somewhere outside of Concourse. It keeps your pipelines portable.

> Side-note; why the reluctance to have integrations directly use the API?

We're focused on our main interface contract, [pipelines](https://concourse-ci.org/pipelines.html)!

The API is not a contract we want to support right now (or potentially ever). Things may look different once we hit everything on our v10 roadmap ([see v10 blog post](https://blog.concourse-ci.org/core-roadmap-towards-v10/)). But for now, the API is not something we want to support. I think not using the API also gives users more control over how they decide to connect Concourse to other systems.

**Comment by jaredstehler** [comment]
Thanks for your input! This definitely makes sense, and simplifies things. I've gone the route of introducing a build registry between concourse and spinnaker.
